test/unified_malloc_framework/memoryPool.hpp

```diff
@@ -84,6 +115,209 @@ TEST_P(umfPoolTest, pow2AlignedAlloc) {
     }
 }
 
+TEST_P(umfPoolTest, freeNullptr) {
+    void *ptr = nullptr;
+    auto ret = umfPoolFree(pool.get(), ptr);
+    ASSERT_EQ(ret, UMF_RESULT_SUCCESS);
+}
+
+TEST_P(umfPoolTest, allocOutOfMem) {
+    static constexpr size_t numProviders = 1;
+    size_t numProvidersRet = 0;
+    std::array<umf_memory_provider_handle_t, numProviders> retProviders;
+
+    auto ret = umfPoolGetMemoryProviders(pool.get(), numProviders,
+                                         retProviders.data(), &numProvidersRet);
+    ASSERT_EQ(ret, UMF_RESULT_SUCCESS);
+    ASSERT_EQ(numProvidersRet, numProviders);
+
+    for (auto provider : retProviders) {
+        if (std::string(umfMemoryProviderGetName(provider)) ==
+            std::string("null")) {
+            GTEST_SKIP();
+        }
+    }
+
+    // test whether memory is kept in a pool accordingly to MaxSize
+    static constexpr size_t allocSize = 16;
+    // MaxSize equals 16 * 1024 * 1024;
+    static constexpr size_t maxAllocs = 1024 * 1024;
+
+    // allocate until oom
+    void *ptr = nullptr;
+    std::vector<void *> allocations;
+
+    for (size_t i = 0; i <= maxAllocs; ++i) {
+        allocations.emplace_back(umfPoolMalloc(pool.get(), allocSize));
+        ASSERT_NE(allocations.back(), nullptr);
+    }
+
+    ASSERT_EQ(umfPoolByPtr(allocations.back()), pool.get());
+
+    // free some memory
+    umfPoolFree(pool.get(), allocations.back());
+
+    allocations.pop_back();
+
+    ptr = umfPoolMalloc(pool.get(), allocSize);
+    ASSERT_NE(ptr, nullptr);
+
+    umfPoolFree(pool.get(), ptr);
+
+    for (auto allocation : allocations) {
+        umfPoolFree(pool.get(), allocation);
+    }
+}
+
+TEST_P(umfPoolTest, multiThreadedMallocFree) {
+    static constexpr size_t allocSize = 64;
+    auto poolMalloc = [](size_t allocSize, umf_memory_pool_handle_t pool) {
+        std::vector<void *> allocations;
+        for (size_t i = 0; i <= 10; ++i) {
+            allocations.emplace_back(umfPoolMalloc(pool, allocSize));
+            ASSERT_NE(allocations.back(), nullptr);
+        }
+
+        for (auto allocation : allocations) {
+            umfPoolFree(pool, allocation);
+        }
+    };
+
+    std::vector<std::thread> threads;
+    for (int i = 0; i < NTHREADS; i++) {
+        threads.push_back(std::thread(poolMalloc, allocSize, pool.get()));
+    }
+
+    for (auto &thread : threads) {
+        thread.join();
+    }
+}
+
+TEST_P(umfPoolTest, multiThreadedpow2AlignedAlloc) {
+#ifdef _WIN32
+    // TODO: implement support for windows
+    GTEST_SKIP();
+#endif
+
+    static constexpr size_t maxAlignment = (1u << 22);
+    static constexpr size_t numAllocs = 4;
+    auto poolPow2AlignedAlloc = [](size_t maxAlignment, size_t numAllocs,
```

Comment:
This looks pretty much the same as in pow2AlignedAlloc. Can you avoid code duplication?